### PR TITLE
Refactor process parsing to accommodate commands with spaces

### LINKED ISSUE #2750: Spaces in command name
According to `man ps`:
> The following user-defined format specifiers may contain spaces: args, cmd, comm, command, fname, ucmd, ucomm, lstart, bsdstart, start.

Therefore `ps` output parsing (using `strings.Fields()`) is broken and is not capable of handling such string:
```
root         107       2 21:34  0.0  0.0     0      0 I<   00:00:00 DWC Notificatio   3 -
```

### LINKED ISSUE #2770: memory cgroup controller disabled by default on some systems
On some systems (such as Raspberry Pi 4) memory cgroup controller is disabled by default. We need to check another controller to determine cgroup path rather than return `"/"`.

### PR DESCRIPTION
**Pull request by iwankgb**
Fixes #2750. Fixes #2770.

### CODE REVIEW

**Comment by Creatone** [comment]
@bobbypage

File: manager/container.go [hunk 1]
```diff
@@ -331,20 +335,28 @@ func (cd *containerData) GetProcessList(cadvisorContainer string, inHostNamespac
...
+		var psr int
+		psr, err = strconv.Atoi(fields[fieldNo])
 		if err != nil {
-			return nil, fmt.Errorf("invalid pid %q: %v", fields[1], err)
+			fieldNo++
+			psr, err = strconv.Atoi(fields[fieldNo])
+			if err != nil {
+				return nil, fmt.Errorf("invalid pid %q: %v", fields[1], err)
+			}
+			cmd += fmt.Sprintf(" %s", fields[11])
```
Comment (bobbypage) [inline manager/container.go:350]: why is this necessary?

File: manager/container.go [hunk 2]
```diff
@@ -331,20 +335,28 @@ func (cd *containerData) GetProcessList(cadvisorContainer string, inHostNamespac
...
 		// convert to bytes
 		vs *= 1024
-		psr, err := strconv.Atoi(fields[11])
+		fieldNo := 11
+		var psr int
+		psr, err = strconv.Atoi(fields[fieldNo])
 		if err != nil {
-			return nil, fmt.Errorf("invalid pid %q: %v", fields[1], err)
+			fieldNo++
+			psr, err = strconv.Atoi(fields[fieldNo])
```
Comment (bobbypage) [inline manager/container.go:346]: If I'm understanding the logic here, the idea is that the cmd should be at index 10 and the psr should be at index 11 ASSUMING there is no spaces in the cmd name (this was old logic).

The fix here is basically assume that cmd==index10 and psr==index11, but if you parse as index11 and get error from `strconv.Atoi` it means you have a space in the cmd name. In that case, it means psr is at some greater index than 11 and keep advancing `fieldNo` until you don't get an error from `strconv.Atoi`.

Am I understanding the workaround here correctly? If so, I feel like this a bit fragile... what if the command name has a space and the string after the space is a number, e.g. "mycmd 1". In that case, wouldn't you assume that "1" actually represents psr value, but actually it's just part of the command name and thus parse the output incorrectly?

File: manager/container_test.go [hunk 3]
```diff
@@ -319,3 +319,19 @@ func TestOnDemandHousekeepingReturnsAfterStopped(t *testing.T) {
...
+	[]byte("root       15886       2 23:51  0.1  0.0     0      0 I    00:00:00 kworker/u8:3-ev   3 -\nroot       15887       2 23:51  0.0  0.0     0      0 I<   00:00:00 kworker/1:2H      1 -\nubuntu     15888    1804 23:51  0.0  0.0  2832  10176 R+   00:00:00 ps                1 8:devices:/user.slice,6:pids:/user.slice/user-1000.slice/session-3.scope,5:blkio:/user.slice,2:cpu,cpuacct:/user.slice,1:na"),
+	[]byte("root         104       2 21:34  0.0  0.0     0      0 I<   00:00:00 kthrotld          3 -\nroot         105       2 21:34  0.0  0.0     0      0 S    00:00:00 irq/41-aerdrv     0 -\nroot         107       2 21:34  0.0  0.0     0      0 I<   00:00:00 DWC Notificatio   3 -\nroot         109       2 21:34  0.0  0.0     0      0 S<   00:00:00 vchiq-slot/0      1 -\nroot         110       2 21:34  0.0  0.0     0      0 S<   00:00:00 vchiq-recy/0      3 -"),
+}
+
+func TestParseProcessList(t *testing.T) {
+	for i, ps := range psOutput {
+		t.Run(fmt.Sprintf("iteration %d", i), func(tt *testing.T) {
+			cd := &containerData{}
+			_, err := cd.parseProcessList("/", true, ps)
+			// checking *only* parsing errors - otherwise /proc would have to be emulated.
```
Comment (bobbypage) [inline manager/container_test.go:333]: do we have some unit tests that check the actual parsing logic and ensure right data is returned after the parse?

**Comment by bobbypage** [comment]
Left a few comments above. BTW, side question (out of scope for this PR): why are we even parsing the results from running `ps`? Wouldn't it make more sense to gather this data from `/proc` or something like that directly?

**Comment by iwankgb** [comment]
I was as surprised as you are. Looking at what I did here makes me think it would be good idea to get rid of output parsing.

File: manager/container_test.go (hunk 3 above)
Comment (iwankgb) [reply manager/container_test.go:333]: Unfortunately not. Loop in `parseProcessList()` besides parsing out put of `ps`:
* reads from cgroup fs
* reads from proc fs

I will try to break it even more so that at least some of the code can be tested.

File: manager/container.go (hunk 1 above)
Comment (iwankgb) [reply manager/container.go]: Because I have forgotten that `strings.Join()` exists :)

File: manager/container.go (hunk 2 above)
Comment (iwankgb) [reply manager/container.go]: I think that parsing logic has been significantly improved and simplified now :)

File: manager/container_test.go (hunk 3 above)
Comment (iwankgb) [reply manager/container_test.go:335]: There is new shiny test function just below :)

**Comment by iwankgb** [comment]
@bobbypage, I hope you will find this code less enigmatic and prone to error :)

File: manager/container.go [hunk 4]
```diff
@@ -283,106 +285,127 @@ func (cd *containerData) getContainerPids(inHostNamespace bool) ([]string, error
...
-		percentCPU, err := strconv.ParseFloat(fields[4], 32)
-		if err != nil {
-			return nil, fmt.Errorf("invalid cpu percent %q: %v", fields[4], err)
-		}
-		percentMem, err := strconv.ParseFloat(fields[5], 32)
-		if err != nil {
-			return nil, fmt.Errorf("invalid memory percent %q: %v", fields[5], err)
-		}
-		rss, err := strconv.ParseUint(fields[6], 0, 64)
+		process, err := cd.parsePsLine(line, cadvisorContainer, inHostNamespace)
```
Comment (bobbypage) [inline manager/container.go:304]: s/process/processInfo (just to be a bit more clear) :)

File: manager/container.go [hunk 5]
```diff
@@ -283,106 +285,127 @@ func (cd *containerData) getContainerPids(inHostNamespace bool) ([]string, error
...
-				RSS:           rss,
-				VirtualSize:   vs,
-				Status:        fields[8],
-				RunningTime:   fields[9],
-				Cmd:           fields[10],
-				CgroupPath:    cgroupPath,
-				FdCount:       fdCount,
-				Psr:           psr,
-			})
+		if cd.isRoot() || cd.info.Name == process.CgroupPath {
```
Comment (bobbypage) [inline manager/container.go:322]: I think the logic here has changed slightly regarding what processInfos are ultimately added to `processes` with respect to the the `cd.info.Name == process.CgroupPath` check

Old logic checked `cd.info.Name == cgroup`
https://github.com/google/cadvisor/blob/3066a8034f2eaf15f69bf1c867df5cd081cd538a/manager/container.go#L364

where `cgroup` was always populated:

https://github.com/google/cadvisor/blob/3066a8034f2eaf15f69bf1c867df5cd081cd538a/manager/container.go#L341

However, in the change here, `process.CgroupPath` will only be populated if the containerData is root:

https://github.com/google/cadvisor/blob/f34a89b2be44fe4915e7f827a4b60dcac9007230/manager/container.go#L393-L394

I think that might be why the old code (very confusingly) had both `cgroup` and `cgroupPath` where cgroupPath was what was set on the v2.ProcessInfo.CGroupPath struct:

https://github.com/google/cadvisor/blob/3066a8034f2eaf15f69bf1c867df5cd081cd538a/manager/container.go#L341-L353

As a result, I think this is currently only including processInfo's that are `isRoot()==true`  which is probably skipping more processInfos than we actually want.

File: manager/container.go [hunk 6]
```diff
@@ -283,106 +285,127 @@ func (cd *containerData) getContainerPids(inHostNamespace bool) ([]string, error
...
+	if cd.isRoot() {
+		info.CgroupPath, err = cd.getCgroupPath(lastTwoFields[1])
+		if err != nil {
+			return nil, fmt.Errorf("could not parse cgroup path from %q: %v", lastTwoFields[1], err)
+		}
+	}
+
+	// Remove the ps command we just ran from cadvisor container.
+	// Not necessary, but makes the cadvisor page look cleaner.
+	if !inHostNamespace && cadvisorContainer == info.CgroupPath && info.Cmd == "ps" {
```
Comment (bobbypage) [inline manager/container.go:402]: ditto re above comment `info.CgroupPath` will only be populated if `cd.isRoot()` unlike previous checke where cgroup was always populated

https://github.com/google/cadvisor/blob/3066a8034f2eaf15f69bf1c867df5cd081cd538a/manager/container.go#L341

https://github.com/google/cadvisor/blob/3066a8034f2eaf15f69bf1c867df5cd081cd538a/manager/container.go#L347

File: manager/container.go [hunk 7]
```diff
@@ -283,106 +285,127 @@ func (cd *containerData) getContainerPids(inHostNamespace bool) ([]string, error
...
+		return nil, fmt.Errorf("invalid virtual size %q: %v", fields[7], err)
+	}
+	// convert to bytes
+	info.RSS *= 1024
+	info.VirtualSize *= 1024
+
+	// According to `man ps`: The following user-defined format specifiers may contain spaces: args, cmd, comm, command,
+	// fname, ucmd, ucomm, lstart, bsdstart, start.
+	// Therefore we need to be able to parse comm that consists of multiple space-separated parts.
+	info.Cmd = strings.Join(fields[10:len(fields)-2], " ")
```
Comment (bobbypage) [inline manager/container.go:385]: good idea re. using `fields[10:len(fields)-2]`as cmd :)

File: manager/container.go [hunk 8]
```diff
@@ -283,106 +285,127 @@ func (cd *containerData) getContainerPids(inHostNamespace bool) ([]string, error
 }
 
 func (cd *containerData) GetProcessList(cadvisorContainer string, inHostNamespace bool) ([]v2.ProcessInfo, error) {
-	// report all processes for root.
-	isRoot := cd.info.Name == "/"
-	rootfs := "/"
-	if !inHostNamespace {
-		rootfs = "/rootfs"
-	}
 	format := "user,pid,ppid,stime,pcpu,pmem,rss,vsz,stat,time,comm,psr,cgroup"
```
Comment (bobbypage) [inline manager/container.go:288]: maybe move this up a constant near `expectedFields`, then it will be clear why `expectedFields = 13`

File: manager/container_test.go (hunk 3 above)
Comment (bobbypage) [reply manager/container_test.go:335]: awesome, thanks for adding tests here!

File: manager/container_test.go [hunk 9]
```diff
@@ -319,3 +321,132 @@ func TestOnDemandHousekeepingReturnsAfterStopped(t *testing.T) {
...
+		nil,
+		&containerData{
+			info: containerInfo{ContainerReference: info.ContainerReference{Name: "/"}},
+		},
+	},
+	{
+		"process with space in name and no cgroup",
+		"root         107       2 21:34  0.0  0.1     3      4 I<   00:20:00 DWC Notificatio   3 -",
+		"/docker/cadvisor",
+		false,
```
Comment (bobbypage) [inline manager/container_test.go:379]: nit: i think it would be a bit more readable if you included the struct names for each field inline here, since it can be a bit confusing what the struct fields map too.

**Comment by bobbypage** [comment]
Thanks for your changes here, I think the updated parsing logic is a lot more clear and less error prone! Left a few smaller comments.

File: manager/container.go (hunk 5 above)
Comment (iwankgb) [reply manager/container.go]: I think I managed to fix the problem that you described. If you take a look at added test cases then you will see that:
* all processes (except `ps` are returned when root container if requested
* only processes belonging to requested container (with no cgroup information set) are returned when it is a non-root container

**Comment by iwankgb** [comment]
@bobbypage, as you can see this PR fixes more than one issue now. I came across #2770 while trying to verify that process are really displayed as expected. They have not been, so I started debugging. I hope it's fine with you - these are still closely related changes to same file.

File: manager/container.go [hunk 10]
```diff
@@ -195,20 +199,24 @@ func (cd *containerData) DerivedStats() (v2.DerivedStats, error) {
...
-	matches := cgroupPathRegExp.FindSubmatch([]byte(cgroups))
+	matches := cgroupMemoryPathRegExp.FindSubmatch([]byte(cgroups))
 	if len(matches) != 2 {
 		klog.V(3).Infof("failed to get memory cgroup path from %q", cgroups)
-		// return root in case of failures - memory hierarchy might not be enabled.
-		return "/", nil
+		// On some systems (e.g. Raspberry PI 4) cgroup memory controlled is disabled by default.
+		matches = cgroupCPUPathRegExp.FindSubmatch([]byte(cgroups))
+		if len(matches) != 2 {
+			// return root in case of failures - memory hierarchy might not be enabled.
```
Comment (bobbypage) [inline manager/container.go:215]: add a klog here if you think this is worth logging? The above klog on 211 might be confusing since it will printed even if we find match by using cpu path regex

File: manager/container_test.go [hunk 11]
```diff
@@ -474,3 +474,51 @@ func TestParsePsLine(t *testing.T) {
 		})
 	}
 }
+
+var cgroupCases = []struct {
+	name    string
```
Comment (bobbypage) [inline manager/container_test.go:479]: can we add case here case where `cgroups==foobar` (some random string that doesn't match any of the regex / cases) and ensure that "/" is returned?

File: manager/container_test.go [hunk 12]
```diff
@@ -474,3 +474,51 @@ func TestParsePsLine(t *testing.T) {
...
+	cgroups string
+	path    string
+}{
+	{
+		"no cgroup",
```
Comment (bobbypage) [inline manager/container_test.go:484]: nit: add field names here and below

File: manager/container.go [hunk 13]
```diff
@@ -48,8 +48,12 @@ var enableLoadReader = flag.Bool("enable_load_reader", false, "Whether to enable
 var HousekeepingInterval = flag.Duration("housekeeping_interval", 1*time.Second, "Interval between container housekeepings")
 
 // cgroup type chosen to fetch the cgroup path of a process.
-// Memory has been chosen, as it is one of the default cgroups that is enabled for most containers.
-var cgroupPathRegExp = regexp.MustCompile(`memory[^:]*:(.*?)[,;$]`)
+// Memory has been chosen, as it is one of the default cgroups that is enabled for most containers...
+var cgroupMemoryPathRegExp = regexp.MustCompile(`memory[^:]*:(.*?)[,;$]`)
```
Comment (bobbypage) [inline manager/container.go:52]: random comment, out of scope for this PR: I wonder if this regex is really needed here. The format seems pretty easily parsable, 

e.g. from my system

```
11:pids:/system.slice/ssh.service,10:devices:/system.slice/ssh.service,4:cpu,cpuacct:/system.slice/ssh.service,2:memory:/system.slice/ssh.service,1:name=systemd:/system.slice/ssh.service,0::/system.slice/ssh.service
```

I think you could just split the string by commas and remove the leading `hierarchy-ID`, e.g. `11:`. We can also see which cgroup controller mounted by reading `/proc/cgroups`. Regex is totally fine for now (and what is used currently) but maybe worth a TODO as future improvement :)

File: manager/container.go [hunk 14]
```diff
@@ -283,106 +288,130 @@ func (cd *containerData) getContainerPids(inHostNamespace bool) ([]string, error
...
+	// Do not report processes from other containers when non-root container requested.
+	if !cd.isRoot() && info.CgroupPath != cd.info.Name {
+		return nil, nil
+	}
+
+	// Remove cgroup information when non-root container requested.
+	if !cd.isRoot() {
+		info.CgroupPath = ""
+	}
+	return &info, nil
```
Comment (bobbypage) [inline manager/container.go:412]: any reason to return pointer here? maybe just return by value, since the return value is dereferenced right away and appended to the processInfo's slice.

File: manager/container.go (hunk 5 above)
Comment (bobbypage) [reply manager/container.go]: Thanks for the fix, your change makes sense!

**Comment by bobbypage** [comment]
Few small comments, but LGTM otherwise. Thanks again for all your work on this!

Can you please squash into maybe two commits, one for the changes of parsing and one for the memory cgroup fix?

File: manager/container.go (hunk 14 above)
Comment (iwankgb) [reply manager/container.go:412]: Yes. I moved logic for deciding whether `ProcessInfo` is of interest to caller: if `nil` is returned then it will not be appended to `processes` slice in `containerData.parseProcessList()`. Such approach keeps logic in one place and, what is even more important, make it testable. I think that `nil` is more meaningful then a zero value `ProcessInfo`.

File: manager/container.go (hunk 10 above)
Comment (iwankgb) [reply manager/container.go:219]: I have completely forgotten about it. Added!

File: manager/container.go (hunk 8 above)
Comment (iwankgb) [reply manager/container.go:295]: Unfortunately these two consts/variables are defined in separate functions.

File: manager/container_test.go (hunk 11 above)
Comment (bobbypage) [reply manager/container_test.go:510]: thanks!

File: manager/container.go (hunk 14 above)
Comment (bobbypage) [reply manager/container.go:420]: got it, that makes sense, thanks!

File: manager/container_test.go (hunk 12 above)
Comment (bobbypage) [reply manager/container_test.go]: thanks

**Comment by bobbypage** [comment]
Thanks again for all your work on this! LGTM!

**Review by bobbypage — APPROVED** [review]
